tensorflow/compiler/tf2tensorrt/convert/trt_optimization_pass.cc

```diff
@@ -241,6 +243,9 @@ Status TRTOptimizationPass::Init(
     TF_RETURN_IF_ERROR(ProfileStrategyFromName(
         params.at("profile_strategy").s(), &profile_strategy_));
   }
+  if (params.count("use_explicit_precision")) {
+    use_explicit_precision_ = params.at("use_explicit_precision").b();
+  }
```

Comment:
You removed the change in the converter API for this. But then how do users pass this parameter to the optimization pass? I was under the impression that we don't need this parameter, instead, the TF-TRT internal will figure out whether we will use_explicit_precision based on the version of the TensorRT. Please correct me if I am not right here.

Reply:
We still need to turn off the constant folding for QDQ ops in grappler

Reply:
My question was: how would you enable or disable use_explicit_precision? Can you describe?

Reply:
Okay I thought we were going to do the enable/disable in follow on PR. But I can update this PR

Reply:
We can do the enable/disable in another PR, but the question is how you are going to enable/disable? 
I don't see why we need to add "use_explicit_precision" to the pass parameter, with the way I have in mind to enable/disable. That is the main reason I am trying to find out how you will enable/disable. Can you describe?

Reply:
Here's the rough call stack for conversion process.

```
  0 API sets up grappler RewriterConfig (either in Python or on new C++ API side)
  1 Run the TRTOptimizationPass::Optimize.
  2 Calls ConvertAfterShapes (segment and outline)
  3     Creates TrtNodeValidator    
  4     Calls segment::SegmentGraph(uses validator)
  5     For each segment:
  6         Calls GetEngineInfo
  7             Calls ConvertSegmentToGraphDef (outlining)
  8     Calls CreateTRTNode
  9         (Updates graph to use TRTEngineOp)
  10         If STATIC: ConvertGraphDefToEngine
  11
  12 TRTEngineOp::ExecuteAsync
  13    ::GetEngine
  14       if(static_engine)
  15          deserialize_engine
  16       if(!static_engine)
  17          ::BuildEngine
  18             convert::ConvertGraphDeftoEngine
  19    ::ExecuteTrtEngine
```

1. At line 0: We need to know explicitly whether the use because we need to turn on the grappler flag to prevent folding emulation for QDQ. In both Python and new C++ API sides, this can be determined by inspecting user's graph and turning on the flag when QDQ is detected and TRT version is >= 8.0.
2. At lines 3, 10, and 18 we need to know to create Validator/Converter respectively.

Option 1:

In order to minimize the number of places where we "detect QDQ" it seems to make sense that we should just add it as a pass parameter that gets set automatically in the C++/Python API functions rather than in any of lines 3-18. The pass parameter would forward this parameter to ConvertAfterShapes, and finally it would be incorporated into the TRTEngineOp, where it is accessible to lines 10 and 18.

Option 2:

We duplicate the "detect QDQ" logic so that Python/C++ API sets the Grappler option `experimental_disable_folding_quantization_emulation` after detecting QDQ, and in addition the function ConvertAfterShapes also does this sort of detection, so that it can pass on the parameter to TRTEngineOp for lines 10 and 18.


I prefer Option 1 since if there is any problem, it is easier to insert a switch to turn off the new style QDQ conversion and use the old style conversion.

What do you think?

Reply:
What would be the problem if we always turning on the grappler flag to prevent folding emulation for QDQ?

Reply:
If you think that's OK, I can do that and infer the parameter at line 3

Reply:
All done. Let me know when you want me  to squash.